### ISSUE TITLE
Edge lines sometimes overlay edge labels

### ISSUE THREAD

**Issue by lounsbrough** [wbkd/react-flow#2494]
### Describe the Bug

In some cases edge lines will overlay the edge labels, which looks bad and can cause readability issues. Whether or not this happens depends on the rendering order of the edges in the final svg.

### Your Example Website or App

https://reactflow.dev/docs/examples/edges/custom-edge/

### Steps to Reproduce the Bug or Issue

1. Go To: https://reactflow.dev/docs/examples/edges/custom-edge/
2. Notice how the `label with styled bg` box is behind the line for `custom edge`

### Expected behavior

As a user, I expect custom edge content to be rendered above the edge lines, so that none of the content in the edges gets hidden, and for better visual appeal.

### Screenshots or Videos

![image](https://user-images.githubusercontent.com/8313853/195445337-52b50101-29a3-4119-ac13-e5a11ee060cc.png)

### Platform

- OS: macOS
- Browser: Chrome
- Version: 106.0.5249.103

### Additional context

There is a clear explanation of svg z-index and rendering order here: https://stackoverflow.com/a/25302276/2373122.

I created a small script that fixes the overlay issue on the example page that I referenced. We can run this in our app, but it's not ideal to do dom surgery every time reactflow renders. Life would be much better if reactflow was careful to render all the edge lines `before` the edges themselves so that the lines do not overlay actual content. With the current structure of the custom edges, I don't know exactly how this would be implemented. I might be missing something here, perhaps there's an easier way to accomplish what we need, but I wanted to pose the issue anyway and see where it goes.

Also worth noting, someone else reported this issue as a discussion, but there was no response there: https://github.com/wbkd/react-flow/discussions/2250. They were on an older version of reactflow, but I am on the current version (v11).

```js
const edgeClass = 'react-flow__edge';
const edgePathClass = 'react-flow__edge-path';

const edges = [...document.querySelectorAll(`.${edgeClass}`)];

if (edges.length === 0) {
  return;
}

const edgesContainer = edges[0].parentElement;

edges.forEach((edge) => {
  const edgeText = edge.querySelector('text');
  const edgePath = edge.querySelector(`.${edgePathClass}`);

  if (!edgeText || !edgePath) {
    return;
  }

  const clonedEdge = edge.cloneNode(true);
  clonedEdge.querySelector(`.${edgePathClass}`).remove();
  edgesContainer.append(clonedEdge);
  edgeText.remove();
});
```

**Comment by moklick** [comment]
Thanks for the detailed report @lounsbrough. I think the best solution for this would be to introduce a dedicated renderer for the edge labels. I will see if and how we could implement something like that. For now, did you try to use the zIndex option for your custom edges? When you use a zIndex, the edge gets rendered in a different SVG container.

**Comment by lounsbrough** [comment]
Thanks for the suggestion @moklick - I will give that a shot. We did end up moving away from the edge labels in favor of just making the edge labels nodes. I will still test out that zIndex prop again and see if I learn anything new to share here.

**Comment by lounsbrough** [comment]
Indeed, the `zIndex` works, but unfortunately it would require us to know which edge labels intersect with paths from other edges. There could even be a recursive case where two edge paths overlap each other's edge labels.

For example, if I set `zIndex` to `2` for id 16 here, and all other edges to `1`, it does indeed render last (above) the other edges, but this doesn't help much in our specific use case:
![image](https://user-images.githubusercontent.com/8313853/195890452-c40ebe03-ce90-40de-90e4-756eda2b2675.png)

For our project we have decided to move forward with using more nodes in place of custom edges, but perhaps fixing this will help someone else in the future.

Thanks!

**Comment by moklick** [comment]
I would love to have a separated div based renderer for the edge labels. I will try to find out if this is possible with the current structure.

**Comment by moklick** [comment]
This is a very first draft, but I could imagine that it could work: https://codesandbox.io/s/portaledgelabelrenderer-ikxqkb?file=/ButtonEdge.js . In this example the edge label is just a div. No `foreignObject` needed. 

We could export a component (`EdgeLabelRenderer`) that one could use as a helper component to render a label to a dedicated edge-label-renderer dom node as seen in the example.

**Comment by moklick** [comment]
This feature is available in the latest release v11.2.0 https://reactflow.dev/docs/api/edges/edge-label-renderer/
